private/bufpkg/bufimage/bufimageutil/image_index.go

```diff
@@ -0,0 +1,107 @@
+// Copyright 2020-2022 Buf Technologies, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package bufimageutil
+
+import (
+	"fmt"
+
+	"github.com/bufbuild/buf/private/bufpkg/bufimage"
+	"github.com/bufbuild/buf/private/pkg/protosource"
+)
+
+type imageIndex struct {
+	NameToDescriptor map[string]protosource.NamedDescriptor
+	NameToExtensions map[string][]protosource.Field
+	NameToOptions    map[string]map[int32]protosource.Field
+}
+
+func newImageIndexForImage(image bufimage.Image) (*imageIndex, error) {
```

Comment:
There's a lot of repetition in this function. We could build a couple functions for adding a type and adding a field to DRY this up and make it a little more readable.

Something along the lines of:

```go
func (i *imageIndex) AddType(namedDescriptor protosource.NamedDescriptor) error {
	if storedDescriptor, ok := i.NameToDescriptor[namedDescriptor.FullName()]; ok && storedDescriptor != namedDescriptor {
		return fmt.Errorf("duplicate for %q: %#v != %#v", namedDescriptor.FullName(), storedDescriptor, namedDescriptor)
	}
	i.NameToDescriptor[namedDescriptor.FullName()] = namedDescriptor
	return nil
}

func (i *imageIndex) AddField(field protosource.Field) error {
	i.NameToDescriptor[field.FullName()] = field
	extendeeName := field.Extendee()
	if isOptionsTypeName(extendeeName) {
		if _, ok := i.NameToOptions[extendeeName]; !ok {
			i.NameToOptions[extendeeName] = make(map[int32]protosource.Field)
		}
		i.NameToOptions[extendeeName][int32(field.Number())] = field
	} else {
		i.NameToExtensions[extendeeName] = append(i.NameToExtensions[extendeeName], field)
	}
	return nil
}

func newImageIndexForImage(image bufimage.Image) (*imageIndex, error) {
	index := &imageIndex{
		NameToDescriptor: make(map[string]protosource.NamedDescriptor),
		NameToExtensions: make(map[string][]protosource.Field),
		NameToOptions:    make(map[string]map[int32]protosource.Field),
	}
	for _, file := range image.Files() {
		protosourceFile, err := protosource.NewFile(newInputFile(file))
		if err != nil {
			return nil, err
		}
		if err := protosource.ForEachMessage(func(message protosource.Message) error {
			index.AddType(message)
			for _, field := range message.Extensions() {
				index.AddField(field)
			}
			return nil
		}, protosourceFile); err != nil {
			return nil, err
		}
		if err = protosource.ForEachEnum(func(enum protosource.Enum) error {
			index.AddType(enum)
			return nil
		}, protosourceFile); err != nil {
			return nil, err
		}
		for _, service := range protosourceFile.Services() {
			index.AddType(service)
		}
		for _, field := range protosourceFile.Extensions() {
			index.AddField(field)
		}
	}
	return index, nil
}
```

There might be problems with the code I included above - it's mostly meant to demonstrate an alternative approach. Take it or leave it.

Reply:
Fair points, I agree that a "builder" would be the right way to go about this. But it's fine as-is - you don't need to roll one out if you don't see a need.